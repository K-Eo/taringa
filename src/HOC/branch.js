import { createFactory } from 'react'

const identity = Component => Component

const branch = (...args) => BaseComponent => {
  const Branch = props => {
    let branchComponent = createFactory(identity(BaseComponent))

    for (let i = 0; i < args.length; i++) {
      let test = args[i][0]

      if (test(props)) {
        let target = args[i][1]
        branchComponent = createFactory(target(BaseComponent))
        break
      }
    }

    return branchComponent(props)
  }

  return Branch
}

export default branch
