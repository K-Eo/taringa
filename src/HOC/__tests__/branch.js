import React from 'react'
import { renderComponent } from 'recompose'

import branch from '../branch'

const BaseComponent = ({ foo }) => {
  return <p>{foo} in BaseComponent</p>
}

const HighComponent = () => {
  return <p>High Component</p>
}

const LowComponent = () => {
  return <p>Low Component</p>
}

describe('branch', () => {
  it('exists', () => {
    expect(branch).toBeDefined()
  })

  it('renders base componente when has no args', () => {
    const Component = branch()(BaseComponent)
    const wrapper = shallow(<Component foo="foo" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders the first truthy branch (first)', () => {
    const Component = branch(
      [props => props.foo === 'foo', renderComponent(HighComponent)],
      [props => props.foo === 'bar', renderComponent(LowComponent)]
    )(BaseComponent)

    const wrapper = shallow(<Component foo="foo" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders the first truthy branch (second)', () => {
    const Component = branch(
      [props => props.foo === 'foo', renderComponent(HighComponent)],
      [props => props.foo === 'bar', renderComponent(LowComponent)]
    )(BaseComponent)

    const wrapper = shallow(<Component foo="bar" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders BaseComponent if none are truthy', () => {
    const Component = branch(
      [props => props.foo === 'foo', renderComponent(HighComponent)],
      [props => props.foo === 'bar', renderComponent(LowComponent)]
    )(BaseComponent)

    const wrapper = shallow(<Component foo="baz" />)

    expect(wrapper).toMatchSnapshot()
  })
})
