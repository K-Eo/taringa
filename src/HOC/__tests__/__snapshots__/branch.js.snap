// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`branch renders BaseComponent if none are truthy 1`] = `
<BaseComponent
  foo="baz"
/>
`;

exports[`branch renders base componente when has no args 1`] = `
<BaseComponent
  foo="foo"
/>
`;

exports[`branch renders the first truthy branch (first) 1`] = `
<renderComponent(HighComponent)
  foo="foo"
/>
`;

exports[`branch renders the first truthy branch (second) 1`] = `
<renderComponent(LowComponent)
  foo="bar"
/>
`;
